components/fxa-client/src/lib.rs

```diff
@@ -85,36 +89,67 @@ pub struct FirefoxAccount {
+    /// [FxaState::Loading] state until then.
     ///
-    ///   - `content_url` - the URL of the Firefox Accounts server to use
-    ///       - For example, use `https://accounts.firefox.com` for the main
-    ///         Mozilla-hosted service.
-    ///   - `client_id` - the registered OAuth client id of the application.
-    ///   - `redirect_uri` - the registered OAuth redirect URI of the application.
-    ///   - `token_server_url_override`: optionally, URL for the user's Sync Tokenserver.
-    ///        - This can be used to support users who self-host their sync data.
-    ///          If `None` then it will default to the Mozilla-hosted Sync server.
-    pub fn new(
-        content_url: &str,
-        client_id: &str,
-        redirect_uri: &str,
-        token_server_url_override: &Option<String>,
-    ) -> FirefoxAccount {
-        FirefoxAccount {
-            internal: std::sync::Mutex::new(internal::FirefoxAccount::new(
-                content_url,
-                client_id,
-                redirect_uri,
-                token_server_url_override.as_deref(),
-            )),
-        }
+    /// Note: because `load_state` runs asynchronously, constructing a `FirefoxAccount` does not
+    /// block on any IO.  This means it's safe to construct a `FirefoxAccount` instance early in
+    /// the startup process.  If applications want to delay loading from disk to speed up the
+    /// startup process, they should should add that delay to `load_state()` rather than waiting to
+    /// construct the `FirefoxAccount` instance.
+    pub fn new(config: FxaConfig, storage: Box<dyn FxaStorage>, event_handler: Box<dyn FxaEventHandler>) -> FirefoxAccount {
+        unimplemented!()
+    }
+
+    /// Get the current state of the client
+    pub fn get_state(&self) -> FxaState {
+        unimplemented!()
     }
 }
 
-uniffi::include_scaffolding!("fxa_client");
+pub enum FxaServer {
+    Release,
+    Stable,
+    Stage,
+    China,
+    LocalDev,
+}
+
+pub struct FxaConfig {
+    /// FxaServer to connect with
+    pub server: FxaServer,
+    /// registered OAuth client id of the application.
+    pub client_id: String,
+    /// `redirect_uri` - the registered OAuth redirect URI of the application.
+    pub redirect_uri: String,
+    ///  URL for the user's Sync Tokenserver. This can be used to support users who self-host their
+    ///  sync data. If `None` then it will default to the Mozilla-hosted Sync server.
+    pub token_server_url_override: Option<String>,
+    /// Device record to register with the FxA server
+    pub device_record: DeviceRecord,
+}
+
+/// Current state of the FxA client
+pub enum FxaState {
+    /// User is logged out
+    Disconnected,
+    /// User is logged in
+    Connected,
+    /// User has been logged out by some external event and needs to re-authenticate, for example a
+    /// password change from another device
+    ReauthenticationNeeded,
```

Comment:
Note: The android-components code makes a formal distinction between "account states" and "progress states".  Account states are the ones above this comment where there's no in-progress operation that could change the state.  Progress states are like the ones below this comment.  I don't love their terminology, but maybe we should split up this enum somehow to reflect that.  The fact that both `Authenticating` and `Loading` have the same warning makes me think this might be a good idea.

Reply:
hrm, that's interesting. Before I saw this comment I was going to suggest that `FxaState::Loading` seemed a bit smelly and that maybe `get_state()` should just return `Option<FxaState>`, with none representing that state. If we did that, then all we have left is "authenticating" - which if we think about it, probably applies to the other states too - eg, `ReauthenticationNeeded` will also need to hit the network and losing that state while we are reauthenticating seems unfortunate. Ditto for disconnection really - we need to hit the network to disconnect - how do we get back to the initial state if those flows fail?

IOW, I think I'm saying that `Authenticating` almost sounds a little like a boolean `is_state_transitioning` or something - the state is still in the *old* state, we are just trying to change it, and probably should reject further attempts to change it while set.

Reply:
I like this -- more information for the consumer seems great in this case.  I updated the code to track 2 things: the current state and if we are in the middle of a transition.  I like the shape of this a lot, although I'm not sure about the type names.